crates/sui-framework/sources/crypto/randomness.move

```diff
@@ -0,0 +1,140 @@
+// // Copyright (c) Mysten Labs, Inc.
+// // SPDX-License-Identifier: Apache-2.0
+
+// TODO: Since the first version is insecure, should we name the module differently? e.g., insecure_randomness?
+
+/// Randomness objects can only be created, set or consumed. They cannot be created and consumed
+/// in the *same* transaction since it might allow validators decide whether to create and use those
+/// objects *after* seeing the randomness they depend on.
+///
+/// - On creation, the object contains the epoch in which it was created and a unique object id.
+///
+/// - After the object creation transaction is committed, anyone can retrieve the BLS signature on
+///   message "randomness":epoch:id from validators (signed using the Threshold-BLS key of that
+///   epoch).
+///
+/// - Anyone that can mutate the object can set the randomness of the object by supplying the BLS
+///   signature. This operation verifies the signature and sets the value of the randomness object
+///   to be the hash of the signature.
+///
+///   Note that there is a single signature that could pass the verification for a specific object,
+///   thus, the only options the owner of the object has after retrieving the signature (and learning
+///   the randomness) is to either set the randomness or leave it unset. Applications that use
+///   Randomness objects must make sure they handle both options (e.g., debit the user on object
+///   creation so even if the user aborts, depending on the randomness it received, the application
+///   is not harmed).
+///
+/// - Once set, the random value can be read/consumed.
+///
+///
+/// This object can be used as a shared-/owned-object.
+///
+module sui::randomness {
+    use std::hash::sha3_256;
+    use std::option;
+    use std::vector;
+    use sui::bcs;
+    use sui::object::{Self, UID, ID};
+    use sui::transfer;
+    use sui::tx_context::{Self, TxContext};
+
+    /// Set is called with an invalid signature.
+    const EInvalidSignature: u64 = 0;
+    /// Already set object cannot be set again.
+    const EAlreadySet: u64 = 1;
+    /// Supplied randomness is not of the right length.
+    const EInvalidRndLength: u64 = 2;
+
+    /// All signatures are prefixed with Domain.
+    const Domain: vector<u8> = b"randomness";
```

Comment:
I'm not sure if this needs to be static or dynamic (probably former) but you can use type of the `T` as a signing parameter through type reflection:
```rust
let unique_type_signature: TypeName = std::type_name::get<T>();
// and then something like
let bytes = sui::bcs::to_bytes(&unique_type_signature);
```